File: app/src/main/java/fr/husta/android/dockersearch/suggestion/SuggestionManager.java
package fr.husta.android.dockersearch.suggestion;

import android.app.Activity;
import android.content.Context;
import android.content.SharedPreferences;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * Store suggestions in {@link SharedPreferences}.
 */
public class SuggestionManager
{

    private static final String PREFS_NAME = "SuggestionPrefs";
    private static final String KEY_SUGGESTIONS = "suggestions";

    private final SharedPreferences sharedPreferences;

    public SuggestionManager(Context context)
    {
        this.sharedPreferences = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
    }

    public SuggestionManager(Activity activity)
    {
        this.sharedPreferences = activity.getPreferences(Context.MODE_PRIVATE);
    }

    /**
     * Save a new suggestion
     */
    public void addSuggestion(String suggestion)
    {
        Set<String> suggestionSet = getSuggestionSet();
        suggestionSet.add(suggestion);
        sharedPreferences.edit().putStringSet(KEY_SUGGESTIONS, suggestionSet).apply();
    }

    /**
     * Remove a specific suggestion
     */
    public void removeSuggestion(String suggestion)
    {
        Set<String> suggestionSet = getSuggestionSet();
        if (suggestionSet.remove(suggestion))
        {
            sharedPreferences.edit().putStringSet(KEY_SUGGESTIONS, suggestionSet).apply();
        }
    }

    /**
     * Get all saved suggestions as a List
     */
    public List<String> getSuggestions()
    {
        return new ArrayList<>(getSuggestionSet());
    }

    /**
     * Clear all stored suggestions
     */
    public void clearSuggestions()
    {
        sharedPreferences.edit().remove(KEY_SUGGESTIONS).apply();
    }

    /**
     * Internal helper to get suggestion set
     */
    private Set<String> getSuggestionSet()
    {
        return new HashSet<>(sharedPreferences.getStringSet(KEY_SUGGESTIONS, new HashSet<>()));
    }
}

File: app/src/main/java/fr/husta/android/dockersearch/suggestion/SuggestionManager.kt
package fr.husta.android.dockersearch.suggestion

import android.app.Activity
import android.content.Context
import android.content.SharedPreferences
import androidx.core.content.edit

/**
 * Store suggestions in [SharedPreferences].
 */
class SuggestionManager {
    private lateinit var sharedPreferences: SharedPreferences

    constructor(context: Context) {
        this.sharedPreferences = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    }

    constructor(activity: Activity) {
        this.sharedPreferences = activity.getPreferences(Context.MODE_PRIVATE)
    }

    // Backing property
    private val _suggestionSet =
        this.sharedPreferences.getStringSet(KEY_SUGGESTIONS, emptySet())
            ?.toMutableSet()
            ?: mutableSetOf()

    // Public read-only view
    val suggestions: Set<String>
        get() = _suggestionSet

    /**
     * Save a new suggestion
     */
    fun addSuggestion(suggestion: String) {
        _suggestionSet.add(suggestion)
        sharedPreferences.edit { putStringSet(KEY_SUGGESTIONS, _suggestionSet) }
    }

    /**
     * Remove a specific suggestion
     */
    fun removeSuggestion(suggestion: String) {
        if (_suggestionSet.remove(suggestion)) {
            sharedPreferences.edit { putStringSet(KEY_SUGGESTIONS, _suggestionSet) }
        }
    }

    /**
     * Clear all stored suggestions
     */
    fun clearSuggestions() {
        sharedPreferences.edit { remove(KEY_SUGGESTIONS) }
    }

    companion object {
        private const val PREFS_NAME = "SuggestionPrefs"
        private const val KEY_SUGGESTIONS = "suggestions"
    }
}
